### PR TITLE
Reuse digest auth challenge

### PR DESCRIPTION
**Pull request by DmitrySamoylov**
Digest auth works this way:
- First we get a 401 with challenge, from which we calculate auth header for the next request
- Next we get 200 with the actual response.

The problem here is that for multiple requests to the same service (let's say to media service `http://192.168.0.128/onvif/Media`), we perform an extra request to obtain a new challenge. However, we can reuse the challenge from the very first 401 and receive 200's in following requests immediately. This challenge gets stale at some point, though, so I handled this as well in this PR.

Before:

```
2024-10-11T11:59:54.728210Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 401 Unauthorized
2024-10-11T11:59:54.763040Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 200 OK
2024-10-11T11:59:54.788537Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 401 Unauthorized
2024-10-11T11:59:54.832126Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 200 OK
2024-10-11T11:59:54.860021Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 401 Unauthorized
2024-10-11T11:59:54.892962Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 200 OK
2024-10-11T11:59:54.919252Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 401 Unauthorized
2024-10-11T11:59:54.953114Z DEBUG onvif::soap::client: http://192.168.0.128/onvif/Media: Response status: 200 OK
```

After:
```
2024-10-11T11:42:00.963856Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 401 Unauthorized
2024-10-11T11:42:00.995125Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:42:01.058180Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:42:01.098189Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:42:01.138092Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:42:01.178275Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
... a minute later nonce gets too old, so we get a new challenge and continue working
2024-10-11T11:43:00.994051Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 401 Unauthorized
2024-10-11T11:43:01.027800Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:43:01.068052Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
2024-10-11T11:43:01.108224Z DEBUG request{uri="http://192.168.0.128/onvif/Media"}: onvif::soap::client: Response status: 200 OK
...
```

For context: we need this for VMSes which tend to have hundreds of cameras connected, and we have to do `GetStreamUri` in a separate request for each camera. ONVIF doesn't have bulk requests, so per-request optimizations now make some sense.

### CODE REVIEW

File: onvif/src/soap/auth/digest.rs [hunk 1]
```diff
@@ -37,29 +41,57 @@ impl Digest {
...
+            // Possible scenarios:
+            // - We've got 401 with a challenge for the first time, we calculate the answer, then
+            //   we get 200 OK. So, a single 401 is never a failure.
+            // - After successful auth the count is 1 because we always store at least one request,
+            //   and the caller decided to reuse the same challenge for multiple requests. But at
+            //   some point, we'll get a 401 with a new challenge and `stale=true`.
+            //   So, we'll get a second 401, and this is also not a failure because after
+            //   calculating the answer to the challenge, we'll get a 200 OK, and will reset the
+            //   counter in `set_success()`.
+            // - Three 401's in a row is certainly a failure.
```
Comment (FSMaxB) [inline onvif/src/soap/auth/digest.rs:78]: This assumption only holds as long as there aren't multiple requests in flight at the same time that would get the same 401 error. Not sure whether that is the case.

File: onvif/src/soap/auth/digest.rs [hunk 2]
```diff
@@ -37,29 +41,57 @@ impl Digest {
...
+    pub fn add_headers(&mut self, mut request: RequestBuilder) -> Result<RequestBuilder, Error> {
+        match &mut self.state {
             State::Default => Ok(request),
-            State::Got401(response) => {
+            State::Got401 { response, count } => {
                 let creds = self.creds.as_ref().ok_or(Error::NoCredentials)?;
 
                 request = request.header("Authorization", digest_auth(response, creds, &self.uri)?);
 
+                *count = count.saturating_add(1);
```
Comment (FSMaxB) [inline onvif/src/soap/auth/digest.rs]: Why does adding the authorization header count as an additional 401?

File: onvif/src/soap/auth/digest.rs (hunk 1 above)
Comment (FSMaxB) [reply onvif/src/soap/auth/digest.rs:78]: Ok, just checked, it holds the `Mutex` from `digest_auth_state`

**Review by FSMaxB — APPROVED** [review]

File: onvif/src/soap/auth/digest.rs (hunk 2 above)
Comment (DmitrySamoylov) [reply onvif/src/soap/auth/digest.rs]: Good catch :+1: Probably some leftovers from my experiments.

**Review by FSMaxB — APPROVED** [review]

**Pull request merged**